packages/sui-react-initial-props/src/withInitialProps.js

```diff
@@ -1,37 +1,110 @@
+  // make a copy of the content safely
+  const windowInitialProps = {...window.__INITIAL_PROPS__}
+  // remove the variable of the window
+  window.__INITIAL_PROPS__ = undefined
+  // return retrieved props from window
+  return windowInitialProps
+}
+
+// extract needed info from props for routeInfo object
+const createRouteInfoFromProps = ({location, params, routes}) => ({
+  location,
+  params,
+  routes
+})
 
 // This HoC creates the PageComponent prepared for the Client Side Rendering
-// It renders a placeholder, if any specified as `renderLoading` on the PageComponent
+// It renders a placeholder, if any specified as `renderLoading` on the PageComponent.
 // Also, executes the getInitialProps to retrieve the initialProps and then
-// Renders the component with the initialProps, hiding the placeholder
-export default ({context, routeInfo}) => Page => {
+// renders the component with the initialProps, hiding the placeholder.
+//
+// Additionally, if `keepMounted` on the PageComponent is true, it keeps the
+// component alive by running getInitialProps again when a route change is done
+// and the very same component is matched by the router, so PageComponent just
+// gets props updated. Also, since PageComponent keeps mounted it will receive
+// an `isLoading` prop while getInitialProps is in progress.
+export default Page => {
+  // gather window initial props for this Page, if present
+  const windowInitialProps = getInitialPropsFromWindow()
+
+  // define Page wrapper component
   const ClientPage = props => {
+    const initialPropsFromWindowRef = useRef(windowInitialProps)
+    // used to know if initialProps has been requested at least once
+    const requestedInitialPropsOnceRef = useRef(!!windowInitialProps)
+    // create routeInfo object from current props which are updated
+    const routeInfo = createRouteInfoFromProps(props)
+    // consume sui context from the context provider
+    const suiContext = useContext(SUIContext)
+    // pathName from context is outdated, so we update it from routeInfo
+    const context = {...suiContext, pathName: routeInfo.location.pathname}
+
     const [{initialProps, isLoading}, setState] = useState({
-      initialProps: {},
-      isLoading: true
+      initialProps: initialPropsFromWindowRef.current || {},
+      isLoading: !initialPropsFromWindowRef.current
     })
 
-    useEffect(function() {
-      Page.getInitialProps({context, routeInfo})
-        .then(initialProps => {
-          setState({initialProps, isLoading: false})
-        })
-        .catch(error => {
-          setState({initialProps: {error}, isLoading: false})
-        })
-    }, [])
+    useEffect(() => {
+      // check if got initial props from window, because then there's no need
+      // to request them again from client
+      if (initialPropsFromWindowRef.current) {
+        initialPropsFromWindowRef.current = undefined
+      } else {
+        Page.getInitialProps({context, routeInfo})
```

Comment:
We moved that useEffect from an only one render when component mounts to a effect that is going to be executed every time `routeInfo.location` changes.
In this case, I think we need to set `isLoading` to `true` when it happens, just before we resolve `getInitialProps` with updated `routeInfo`.

```suggestion
        setState({initialProps, isLoading: true})

        Page.getInitialProps({context, routeInfo})
```